rs/nns/sns-wasm/src/sns_wasm.rs

```diff
@@ -422,12 +423,30 @@ where
         if hash != wasm.sha256_hash() {
             return AddWasmResponse {
                 result: Some(add_wasm_response::Result::Error(SnsWasmError {
-                    message: format!("Invalid Sha256 given for submitted WASM bytes. Provided hash was '{}'  but calculated hash was '{}'",
-                                   hash_to_hex_string(&hash), wasm.sha256_string())
-                }))
+                    message: format!(
+                        "Invalid Sha256 given for submitted WASM bytes. Provided hash was '{}' \
+                         but calculated hash was '{}'",
+                        hash_to_hex_string(&hash),
+                        wasm.sha256_string()
+                    ),
+                })),
             };
         }
 
+        let metadata = match Self::read_wasm_metadata_or_err(&wasm) {
+            Ok(metadata) => metadata,
+            Err(err) => {
+                println!("err = {}, wasm = `{:?}`", err, wasm);
+                return AddWasmResponse {
+                    result: Some(add_wasm_response::Result::Error(SnsWasmError {
+                        message: format!("Cannot read metadata sections from WASM: {}", err),
+                    })),
+                };
+            }
+        };
+
+        let metadata = metadata.into_iter().map(MetadataSectionPb::from).collect();
```

Comment:
At this point, we need to decide what to do if the metadata is too big.

Any suggestions how to best restrict it?

If we don't enforce any limits here, in the worst case, storing the metadata in SNS-W would increase the memory consumption by almost 2X.